extension/src/time_series/pipeline/fill_to.rs

```diff
@@ -0,0 +1,248 @@
+
+type Interval = pg_sys::Datum;
+
+// TODO: there are one or two other gapfill objects in this extension, these should be unified
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug, FlatSerializable)]
+#[repr(u64)]
+pub enum FillToMethod {
+    LOCF,
+    Interpolate,
+    Nearest,
+}
+
+impl FillToMethod {
+    pub fn fill_point(&self, lhs: &TSPoint, rhs: &TSPoint, target_ts: i64) -> TSPoint {
+        match *self {
+            FillToMethod::LOCF => TSPoint{ts: target_ts, val: lhs.val},
+            FillToMethod::Interpolate => {
+                let interval = rhs.ts as f64 - lhs.ts as f64;
+                let left_wt = 1. - (target_ts - lhs.ts) as f64 / interval;
+                let right_wt = 1. - (rhs.ts - target_ts) as f64 / interval;
+                TSPoint{ts: target_ts, val: lhs.val * left_wt + rhs.val * right_wt}
+            },
+            FillToMethod::Nearest =>
+                if rhs.ts - target_ts >= target_ts - lhs.ts {
+                    TSPoint{ts: target_ts, val: lhs.val}
+                } else {
+                    TSPoint{ts: target_ts, val: rhs.val}
+                },
+        }
+    }
+}
+
+// TODO is (immutable, parallel_safe) correct?
+#[pg_extern(
+    immutable,
+    parallel_safe,
+    name="fill_to",
+    schema="toolkit_experimental"
+)]
+pub fn fillto_pipeline_element<'e> (
+    fill_method: String,
+    interval: Interval,
+) -> toolkit_experimental::UnstableTimevectorPipeline<'e> {
+    unsafe {
+        let interval = interval as *const pg_sys::Interval;
+        if (*interval).day > 0 || (*interval).month > 0 {
+            panic!("downsample intervals are currently restricted to stable units (hours or smaller)");
+        }
+        let interval = (*interval).time;
+
+        let fill_method = match fill_method.to_lowercase().as_str() {
+            "locf" => FillToMethod::LOCF,
+            "interpolate" => FillToMethod::Interpolate,
+            "linear" => FillToMethod::Interpolate,
+            "nearest" => FillToMethod::Nearest,
+            _ => panic!("Invalid fill method")
+        };
+
+        Element::FillTo {
+            interval,
+            fill_method
+        }.flatten()
+    }
+}
+
+pub fn fill_to<'s>(
+    series: &toolkit_experimental::Timevector<'s>,
+    element: &toolkit_experimental::Element
+) -> toolkit_experimental::Timevector<'s> {
+    let (interval, method) = match element {
+        Element::FillTo{interval, fill_method} => (*interval, fill_method),
+        _ => unreachable!()
+    };
+
+    match series.series {
+        SeriesType::ExplicitSeries{..} => panic!("Timeseries must be sorted prior to passing to fill_to"),
+        SeriesType::NormalSeries{step_interval, ..} => if step_interval <= interval { return series.clone(); },
+        _ => ()
+    }
+
+    let mut result: Option<Vec<TSPoint>> = None;
+    let mut it = series.iter().peekable();
+    let mut current = it.next();
+
+    while current.is_some() {
+        let lhs = current.unwrap();
```

Comment:
```suggestion
    while let (lhs) = current {
```

Reply:
nice, knew there had to be a better way to do this.